distros/I/IPChains/documentation/IPChains.3
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH IPChains 3 "perl 5.004, patch 04" "23/Feb/99" "User Contributed Perl Documentation"
.UC
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
IPChains \- Create and Manipulate ipchains via Perl
.SH "SYNOPSIS"
use IPChains;
.PP
$fw = IPChains->\fInew\fR\|(\-option => value, ... );
\f(CW$fw\fR\->\fIappend\fR\|('chain');
.SH "DESCRIPTION"
This module acts as an interface to the \fIipchains\fR\|(8) userspace utility by
Paul \*(L"Rusty\*(R" Russell (http://www.rustcorp.com/linux/ipchains/). It attempts 
to include all the functionality of the original code with a simplified 
user interface via Perl. In addition, plans for log parsing facilities, an
integrated interface to ipmasqadm, and possibly traffic shaping are slated
for up and coming versions. 
.PP
The \fInew()\fR and \fIattribute()\fR methods support the following options:
.Ip "\fBSource\fR" 4
Specifies origination address of packet. Appending hostmask to this address using a / is 
\s-1OK\s0, as well as specifying it separately (see SourceMask).
.Ip "\fBSourceMask\fR" 4
Hostmask for origination address. Can either be in 24 or 255.255.255.0 style. 
.Ip "\fBSourcePort\fR" 4
Specific port or port range (use xxx:xxx to denote range), requires specific 
protocol specification.
.Ip "\fBDest\fR" 4
Specifies destination address of packet. Appending hostmask to this address using a / is
\s-1OK\s0, as well as specifying it separately (see DestMask)
.Ip "\fBDestMask\fR" 4
Destination address, (see SourceMask).
.Ip "\fBDestPort\fR" 4
Destination Port, (see SourcePort).
.Ip "\fBProt\fR" 4
Protocol. Can be tcp, udp, icmp, or all. Required for specifying specific \fIport\fR\|(s).
.Ip "\fB\s-1ICMP\s0\fR" 4
\s-1ICMP\s0 Name/Code (in place of port when \s-1ICMP\s0 is specified as protocol).
.Sp
Here is a small table of some of the most common \s-1ICMP\s0 packets:
.Sp
.Vb 1
\&       Number  Name                     Required by
.Ve
.Vb 6
\&       0       echo-reply               ping
\&       3       destination-unreachable  Any TCP/UDP traffic.
\&       5       redirect                 routing if not running 
\&                                        routing daemon
\&       8       echo-request             ping
\&       11      time-exceeded            traceroute
.Ve
.Ip "\fBRule\fR" 4
Target. Can be \s-1ACCEPT\s0, \s-1DENY\s0, \s-1REJECT\s0, \s-1MASQ\s0, \s-1REDIRECT\s0, \s-1RETURN\s0, or a user-defined chain.
Note: This is case sensitive.
.Ip "\fBInterface\fR" 4
Specify a specify interface as part of the criteria (ie, eth0, ppp0, etc.).
.Ip "\fBFragment\fR" 4
Rule only refers to second and further fragments of fragmented packets (1 or 0).
.Ip "\fBBidir\fR" 4
Makes criteria effective in both directions (1 or 0).
.Ip "\fBVerbose\fR" 4
Set verbose option for setting rules or \fIlist()\fR (1 or 0).
.Ip "\fBNumeric\fR" 4
Show output from \fIlist()\fR in numeric format. No \s-1DNS\s0 lookups, etc.. (1 or 0).
.Ip "\fBLog\fR" 4
Enable kernel logging (via syslog, kern.info) of matched packets (1 or 0).
.Ip "\fBOutput\fR" 4
Copy matching packets to the userspace device (advanced).
.Ip "\fBMark\fR" 4
Mark matching packets with specified number (advanced).
.Ip "\fB\s-1TOS\s0\fR" 4
Used for modifying the \s-1TOS\s0 field in the \s-1IP\s0 header. Takes 2 args, \s-1AND\s0 and \s-1XOR\s0 masks,
(ie, (\s-1TOS\s0 => ["0x01\*(R", \*(L"0x10"])). This feature is highly untested.
.Sp
The first mask is ANDed with the packet's current \s-1TOS\s0, and the
second mask is XORed with it. Use the following table for reference:
.Sp
.Vb 1
\&       TOS Name                Value           Typical Uses
.Ve
.Vb 5
\&       Minimum Delay           0x01 0x10       ftp, telnet
\&       Maximum Throughput      0x01 0x08       ftp-data
\&       Maximum Reliability     0x01 0x04       snmp
\&       Minimum Cost            0x01 0x02       nntp
\& 
.Ve
.Ip "\fBExact\fR" 4
Display exact numbers in byte counters instead of
numbers rounded in K's, M's, or G's (1 or 0).
.Ip "\fB\s-1SYN\s0\fR" 4
Only match \s-1TCP\s0 packets with the \s-1SYN\s0 bit set and the 
\s-1ACK\s0 and \s-1FIN\s0 bits cleared (1 or 0).
.SH "METHODS"
The following methods are available to you:
.Ip "\fBnew()\fR" 4
$fw = IPChains->\fInew\fR\|(option => value, ...) create new fw object with 
options
.Ip "\fBattribute()\fR" 4
$fw->\fIattribute\fR\|(option, value) to set option to value, \s-1OR\s0
\f(CW$value\fR = \f(CW$obj\fR\->\fIattribute\fR\|(option) to get current value of option.
.Ip "\fBclopts()\fR" 4
$fw->\fIclopts()\fR clears all option settings (do this before
calling methods like \fIlist()\fR, \fIflush()\fR, \fIdelete()\fR, etc. that
take only a few specific options).
.Ip "\fBappend()\fR" 4
$fw->\fIappend\fR\|(chain) appends current rule to end of chain
.Ip "\fBinsert()\fR" 4
$fw->\fIinsert\fR\|(chain, rulenum) inserts rule at position rulenum
in chain. If rulenum is omitted 1 is assumed.
.Ip "\fBreplace()\fR" 4
$fw->\fIreplace\fR\|(chain, rulenum) replace rule at rulenum in chain with 
current rule.
.Ip "\fBdelete()\fR" 4
$fw->\fIdelete\fR\|(chain, rulenum) deletes rule rulenum from chain.
.Ip "\fBcheck()\fR" 4
$fw->\fIcheck\fR\|(chain) check given packet against chain for testing.
.Ip "\fBflush()\fR" 4
$fw->\fIflush\fR\|(chain) deletes all rules from chain.
.Ip "\fBlist()\fR" 4
$fw->\fIlist\fR\|(chain) lists all rules defined for chain.
.Ip "\fBzero()\fR" 4
$fw->\fIzero\fR\|(chain) zero's all packet counters for chain. Cannot zero 
counters for chain policy.
.Ip "\fBmasq()\fR" 4
$fw->\fImasq()\fR lists current masqueraded connections.
.Ip "\fBnew_chain()\fR" 4
$fw->\fInew_chain\fR\|(chain) creates new user defined chain.
.Ip "\fBdel_chain()\fR" 4
$fw->\fIdel_chain\fR\|(chain) delete user defined chain.
.Ip "\fBset_policy()\fR" 4
$fw->\fIset_policy\fR\|(chain) set default policy for chain. Takes
Rule option only.
.SH "EXAMPLES"
To set the default policy for the \*(L"forward\*(R" chain to DENY:
.PP
.Vb 1
\& use IPChains;
.Ve
.Vb 2
\& $fw = IPChains->new(Rule => "DENY");
\& $fw->set_policy("forward");
.Ve
To list current rules in \*(L"input\*(R" chain to stdout (without parsing through /proc/net/ip_fw*):
.PP
.Vb 1
\& use IPChains;
.Ve
.Vb 2
\& $fw = IPChains->new(Verbose => 1);
\& $fw->list("input");
.Ve
To create a rule that would allow all traffic on an internal lan, and deny
all tcp traffic from external hosts on relevant ports, and log it,you could 
use something like:
.PP
.Vb 1
\& use IPChains;
.Ve
.Vb 9
\& $internal = IPChains->new(Source    => "192.168.100.0/24",
\&                           Rule      => "ACCEPT",
\&                           Interface => "eth0");
\& $external = IPChains->new(Interface => "ppp0",
\&                           Prot      => "tcp",
\&                           DestPort  => "0:1024",
\&                           Log       => 1);
\& $internal->append("input");
\& $external->append("input");
.Ve
You could also create one object, set up the attributes, \fIappend()\fR it, then
use \fIclopts()\fR to clear it's options, then use \fIattribute()\fR to individually
specify it's next set of options, then \fIappend()\fR it again with the new rule.
See the examples/ subdirectory in the IPChains.pm source for more examples.
.SH "BUGS"
Much of this is highly untested. Masquerading timeout setting and negative
attributes (!) aren't yet implemented.
Much of what's planned to be done hasn't been yet. This is
to be considered nothing more than an early beta to work out bugs in the
basic code, and get feedback on usefulness and improvements that could
be made.
.SH "AUTHOR"
Jessica Hutchison (j@splatlabs.com). Please feel free to email me with
feedback, questions, or comments (or indeed patches/additions).
.SH "COPYRIGHT"
This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself, with the exception
of the libipfwc.c, ipchains.c, and the files in include/ which 
have separate terms derived from those of the original ipchains 
sources. See COPYING for details of this license. Please see
README.ipchains for the README that was included with the
original source code for ipchains and contains copyrights and
credits for such. 

.rn }` ''
.IX Title "IPChains 3"
.IX Name "IPChains - Create and Manipulate ipchains via Perl"

.IX Header "NAME"

.IX Header "SYNOPSIS"

.IX Header "DESCRIPTION"

.IX Item "\fBSource\fR"

.IX Item "\fBSourceMask\fR"

.IX Item "\fBSourcePort\fR"

.IX Item "\fBDest\fR"

.IX Item "\fBDestMask\fR"

.IX Item "\fBDestPort\fR"

.IX Item "\fBProt\fR"

.IX Item "\fB\s-1ICMP\s0\fR"

.IX Item "\fBRule\fR"

.IX Item "\fBInterface\fR"

.IX Item "\fBFragment\fR"

.IX Item "\fBBidir\fR"

.IX Item "\fBVerbose\fR"

.IX Item "\fBNumeric\fR"

.IX Item "\fBLog\fR"

.IX Item "\fBOutput\fR"

.IX Item "\fBMark\fR"

.IX Item "\fB\s-1TOS\s0\fR"

.IX Item "\fBExact\fR"

.IX Item "\fB\s-1SYN\s0\fR"

.IX Header "METHODS"

.IX Item "\fBnew()\fR"

.IX Item "\fBattribute()\fR"

.IX Item "\fBclopts()\fR"

.IX Item "\fBappend()\fR"

.IX Item "\fBinsert()\fR"

.IX Item "\fBreplace()\fR"

.IX Item "\fBdelete()\fR"

.IX Item "\fBcheck()\fR"

.IX Item "\fBflush()\fR"

.IX Item "\fBlist()\fR"

.IX Item "\fBzero()\fR"

.IX Item "\fBmasq()\fR"

.IX Item "\fBnew_chain()\fR"

.IX Item "\fBdel_chain()\fR"

.IX Item "\fBset_policy()\fR"

.IX Header "EXAMPLES"

.IX Header "BUGS"

.IX Header "AUTHOR"

.IX Header "COPYRIGHT"



distros/I/IPChains/IPChains.pm
package IPChains;
require Exporter;
require DynaLoader;
@ISA = qw(Exporter DynaLoader);
package IPChainsc;
bootstrap IPChains;
var_IPChains_init();
@EXPORT = qw();
$VERSION = 0.6;

# ---------- BASE METHODS -------------

package IPChains;

sub TIEHASH {
    my ($classname,$obj) = @_;
    return bless $obj, $classname;
}

sub CLEAR { }

sub this {
    my $ptr = shift;
    return tied(%$ptr);
}


# ------- FUNCTION WRAPPERS --------

package IPChains;

*ipfw_init = *IPChainsc::ipfw_init;
*ipfw_append = *IPChainsc::ipfw_append;
*ipfw_delete = *IPChainsc::ipfw_delete;
*ipfw_check = *IPChainsc::ipfw_check;
*ipfw_replace = *IPChainsc::ipfw_replace;
*ipfw_insert = *IPChainsc::ipfw_insert;
*ipfw_list = *IPChainsc::ipfw_list;
*ipfw_flush = *IPChainsc::ipfw_flush;
*ipfw_zero = *IPChainsc::ipfw_zero;
*ipfw_masq = *IPChainsc::ipfw_masq;
*ipfw_new_chain = *IPChainsc::ipfw_new_chain;
*ipfw_del_chain = *IPChainsc::ipfw_del_chain;
*ipfw_set_policy = *IPChainsc::ipfw_set_policy;

# -- End SWIG Generated Code -------------

package IPChains;

sub new {
	my $self = shift;
	my %args = @_;

	# need to verify TOS and add -S and ! support eventually.

	bless {
		"Source"	=>	$args{Source},		# accepts !
		"Dest"		=>	$args{Dest},		# accepts !
		"SourceMask"	=>	$args{SourceMask},
		"DestMask"	=>	$args{DestMask},
		"SourcePort"	=>	$args{SourcePort},	# accepts !
		"DestPort"	=>	$args{DestPort},	# accepts !
		"Prot"		=>	$args{Prot},		# required for port(s)
		"ICMP"		=>	$args{ICMP},		# no code with -d & no ports.
		"Rule"		=>	$args{Rule},		# -j in ipchains
		"Interface"	=>	$args{Interface},	# accepts !
		"Fragment"	=>	$args{Fragment},	# bool, accepts !
		"Bidir"		=>	$args{Bidir},		# bool
		"Verbose"	=>	$args{Verbose},		# bool
		"Numeric"	=>	$args{Numeric},		# bool - for L
		"Log"		=>	$args{Log},		# bool
		"Output"	=>	$args{Output}, 		# takes "maxsize" arg
		"Mark"		=>	$args{Mark},		# takes "mark value"
		"Exact"		=>	$args{Exact},		# bool - for L
		"SYN"		=>	$args{SYN},		# bool, accepts !
		"TOS"		=>	$args{TOS}
	        "RedirectPort"  =>      $args{RedirectPort}     # redirect port
		}, $self;
}

sub attribute {
	my ( $self, $attrib, $value ) = @_;
	if ( @_ == 3 ) {
		$self->{$attrib} = $value;
	} else {
		$value = $self->{$attrib};
	}
	return $value;
}

sub clopts {
	my ( $self ) = shift;
	%$self = ();
}

sub set_opts {
	my $self = shift;
	my @opts = ();

	# Program name needs to be @opts[0] to work.
	push(@opts, "IPChains");

	# Protocol
	if (defined($self->{Prot})) {
		push(@opts, "-p", "$self->{Prot}");
	}
	# Source address, mask, and port
	if (defined($self->{Source}) && ! defined($self->{SourceMask})) {
		push(@opts, "-s", "$self->{Source}");
	} elsif (defined($self->{Source}) && defined ($self->{SourceMask})) {
		push(@opts, "-s", "$self->{Source}/$self->{SourceMask}");
	}
	if (defined($self->{SourcePort}) && defined($self->{ICMP})) {
		die "Cannot Specify Port and ICMP type\n";
	}
	if (defined($self->{SourcePort}) && ! defined($self->{Source})) {
	  push(@opts, "-s", "0/0", "$self->{SourcePort}");
	}
	if (defined($self->{SourcePort}) && defined($self->{Source})) {
		push(@opts, "$self->{SourcePort}");
	}
	# Dest address, mask, and port
	if (defined($self->{Dest}) && ! defined($self->{DestMask})) {
		push(@opts, "-d", "$self->{Dest}");
	} elsif (defined($self->{Dest}) && defined ($self->{DestMask})) {
		push(@opts, "-d", "$self->{Dest}/$self->{DestMask}");
	}
	if (defined($self->{DestPort}) && defined($self->{ICMP})) {
		die "Cannot Specify Port and ICMP type\n";
	}
	if (defined($self->{DestPort}) && ! defined($self->{Dest})) {
	  push(@opts, "-d", "0/0", "$self->{DestPort}");
	}
	if (defined($self->{DestPort}) && defined($self->{Dest})) {
		push(@opts, "$self->{DestPort}");
	}
	if ( defined($self->{ICMP} )) {
		push @opts, "--icmp-type", $self->{ICMP};
	}
	    
	# "Jump" rule
	if (defined($self->{Rule})) {
		if (($self->{Rule} eq "REDIRECT" ) and (defined($self->{RedirectPort}))) {
			push(@opts, "-j", "$self->{Rule}", "$self->{RedirectPort}");
		} else {	 
			push(@opts, "-j", "$self->{Rule}");
		}	
	}
	if (defined($self->{Interface})) {
		push(@opts, "-i", "$self->{Interface}");
	}
	if ($self->{Fragment}) { push(@opts, "-f"); }
	if ($self->{Bidir}) { push(@opts, "-b"); }
	if ($self->{Verbose}) {	push(@opts, "-v"); }
	if ($self->{Numeric}) {	push(@opts, "-n"); }
	if ($self->{Log}) { push(@opts, "-l"); }
	if (defined($self->{Output})) {
		if ($self->{Output} == 1) {
			push(@opts, "-o");	
		} else {
			push(@opts, "-o", "$self->{Output}");
		}
	}
	if (defined($self->{Mark})) { 
		push(@opts, "-m", "$self->{Mark}"); 
	}
	if ($self->{Exact}) { push(@opts, "-x"); }
	if (defined($self->{SYN})) {
		if ("$self->{SYN}" =~ /\!/) {
			push(@opts, "!", "-y");
		} else {
			push(@opts, "-y");

		}
	}
	# UNTESTED. MAY SPONTAINIOUSLY COMBUST.
	if (defined($self->{TOS})) {
		push(@opts, "-t", "$self->{TOS}->[0]", "$self->{TOS}->[1]");
	}


	return \@opts;
}


sub append {
	my ($self, $chain) = @_;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_append($argc, $opts, $chain);
}	
	
sub insert {
	my ($self, $chain, $rulenum) = @_;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_insert($argc, $opts, $chain, $rulenum);
}

sub replace {
	my ($self, $chain, $rulenum) = @_;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_replace($argc, $opts, $chain, $rulenum);
}

sub delete {
	my ($self, $chain, $rulenum) = @_;
	$opts = set_opts($self);
        $argc = @$opts;
	ipfw_delete($argc, $opts, $chain, $rulenum);
}

sub check {
	my ($self, $chain) = @_;
        $opts = set_opts($self);
        $argc = @$opts;
        ipfw_check($argc, $opts, $chain);
}

sub flush {
	my ($self, $chain) = @_;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_flush($argc, $opts, $chain);
}

sub list {
	my ($self, $chain) = @_;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_list($argc, $opts, $chain);
}

# list_chains() contributed by Sven Koch <haegar@comunit.net>
sub list_chains {
	#my ($self) = @_;
	if (open(LIST, "</proc/net/ip_fwnames"))
	{
		my @result = grep { chomp; s/^([\w-]+).*/$1/; !/^(input|output|forward)$/ } <LIST>;
		close(LIST);
		return @result;
	} else {
		return undef;
	}
}

sub zero {
	my ($self, $chain) = @_;
	@opts = qw(IPChains $chain);
	$argc = @opts;
	ipfw_zero($argc, \@opts);
}

sub new_chain { 
	my ($self, $chain) = @_;
	@opts = qw(IPChains);
	$argc = @opts;
	ipfw_new_chain($argc, \@opts, $chain);
}

sub del_chain {
	my ($self, $chain) = @_;
	@opts = qw(IPChains);
	$argc = @opts;
	ipfw_del_chain($argc, \@opts, $chain);
}

sub set_policy {
	my ($self, $chain) = @_;
	my @opts;
	if (!defined($self->{Rule})) {	
		print "set_policy requires policy.\n";
	} else {
		push(@opts, "IPChains","$self->{Rule}");
	}
	$argc = @opts;
	ipfw_set_policy($argc, \@opts, $chain);

}

sub masq {
	my ($self) = shift;
	$opts = set_opts($self);
	$argc = @$opts;
	ipfw_masq($argc, $opts);
}


1;

__END__

=head1 NAME

IPChains - Create and Manipulate ipchains via Perl

=head1 SYNOPSIS

use IPChains;

$fw = IPChains->new(-option => value, ... );
$fw->append('chain');

=head1 DESCRIPTION

This module acts as an interface to the ipchains(8) userspace utility by
Paul "Rusty" Russell (http://www.rustcorp.com/linux/ipchains/). It attempts 
to include all the functionality of the original code with a simplified 
user interface via Perl. In addition, plans for log parsing facilities, an
integrated interface to ipmasqadm, and possibly traffic shaping are slated
for up and coming versions. 

The new() and attribute() methods support the following options:

=over 4

=item B<Source>

Specifies origination address of packet. Appending hostmask to this address using a / is 
OK, as well as specifying it separately (see SourceMask).

=item B<SourceMask>

Hostmask for origination address. Can either be in 24 or 255.255.255.0 style. 

=item B<SourcePort>

Specific port or port range (use xxx:xxx to denote range), requires specific 
protocol specification.

=item B<Dest>

Specifies destination address of packet. Appending hostmask to this address using a / is
OK, as well as specifying it separately (see DestMask)

=item B<DestMask>

Destination address, (see SourceMask).

=item B<DestPort>

Destination Port, (see SourcePort).

=item B<Prot>

Protocol. Can be tcp, udp, icmp, or all. Required for specifying specific port(s).

=item B<ICMP>

ICMP Name/Code (in place of port when ICMP is specified as protocol).

Here is a small table of some of the most common ICMP packets:

       Number  Name                     Required by

       0       echo-reply               ping
       3       destination-unreachable  Any TCP/UDP traffic.
       5       redirect                 routing if not running 
                                        routing daemon
       8       echo-request             ping
       11      time-exceeded            traceroute


=item B<Rule>

Target. Can be ACCEPT, DENY, REJECT, MASQ, REDIRECT, RETURN, or a user-defined chain.
Note: This is case sensitive.

=item B<Interface>

Specify a specify interface as part of the criteria (ie, eth0, ppp0, etc.).

=item B<Fragment>

Rule only refers to second and further fragments of fragmented packets (1 or 0).

=item B<Bidir>

Makes criteria effective in both directions (1 or 0).

=item B<Verbose>

Set verbose option for setting rules or list() (1 or 0).

=item B<Numeric>

Show output from list() in numeric format. No DNS lookups, etc.. (1 or 0).

=item B<Log>


Enable kernel logging (via syslog, kern.info) of matched packets (1 or 0).

=item B<Output>

Copy matching packets to the userspace device (advanced).

=item B<Mark>

Mark matching packets with specified number (advanced).

=item B<TOS>


Used for modifying the TOS field in the IP header. Takes 2 args, AND and XOR masks,
(ie, (TOS => ["0x01", "0x10"])). This feature is highly untested.

The first mask is ANDed with the packet's current TOS, and the
second mask is XORed with it. Use the following table for reference:

       TOS Name                Value           Typical Uses

       Minimum Delay           0x01 0x10       ftp, telnet
       Maximum Throughput      0x01 0x08       ftp-data
       Maximum Reliability     0x01 0x04       snmp
       Minimum Cost            0x01 0x02       nntp
 

=item B<Exact>

Display exact numbers in byte counters instead of
numbers rounded in K's, M's, or G's (1 or 0).

=item B<SYN>

Only match TCP packets with the SYN bit set and the 
ACK and FIN bits cleared (1 or 0).

=back

=head1 METHODS

The following methods are available to you:

=over 4

=item B<new()>

$fw = IPChains->new(option => value, ...) create new fw object with 
options

=item B<attribute()>

$fw->attribute(option, value) to set option to value, OR
$value = $obj->attribute(option) to get current value of option.

=item B<clopts()>

$fw->clopts() clears all option settings (do this before
calling methods like list(), flush(), delete(), etc. that
take only a few specific options).

=item B<append()>

$fw->append(chain) appends current rule to end of chain

=item B<insert()>

$fw->insert(chain, rulenum) inserts rule at position rulenum
in chain. If rulenum is omitted 1 is assumed.

=item B<replace()>

$fw->replace(chain, rulenum) replace rule at rulenum in chain with 
current rule.

=item B<delete()>

$fw->delete(chain, rulenum) deletes rule rulenum from chain.

=item B<check()>

$fw->check(chain) check given packet against chain for testing.

=item B<flush()>

$fw->flush(chain) deletes all rules from chain.

=item B<list()>

$fw->list(chain) lists all rules defined for chain.

=item B<list_chains()>

$fw->list_chains() returns array with the names of all user-defined chains
or undef if none exist.

=item B<zero()>

$fw->zero(chain) zero's all packet counters for chain. Cannot zero 
counters for chain policy.

=item B<masq()>

$fw->masq() lists current masqueraded connections.

=item B<new_chain()>

$fw->new_chain(chain) creates new user defined chain.

=item B<del_chain()>

$fw->del_chain(chain) delete user defined chain.

=item B<set_policy()>

$fw->set_policy(chain) set default policy for chain. Takes
Rule option only.

=back

=head1 EXAMPLES

To set the default policy for the "forward" chain to DENY:

 use IPChains;

 $fw = IPChains->new(Rule => "DENY");
 $fw->set_policy("forward");

To list current rules in "input" chain to stdout (without parsing through /proc/net/ip_fw*):

 use IPChains;

 $fw = IPChains->new(Verbose => 1);
 $fw->list("input");

To create a rule that would allow all traffic on an internal lan, and deny
all tcp traffic from external hosts on relevant ports, and log it,you could 
use something like:

 use IPChains;

 $internal = IPChains->new(Source    => "192.168.100.0/24",
                           Rule      => "ACCEPT",
                           Interface => "eth0");
 $external = IPChains->new(Interface => "ppp0",
                           Prot      => "tcp",
                           DestPort  => "0:1024",
                           Log       => 1);
 $internal->append("input");
 $external->append("input");

You could also create one object, set up the attributes, append() it, then
use clopts() to clear it's options, then use attribute() to individually
specify it's next set of options, then append() it again with the new rule.
See the examples/ subdirectory in the IPChains.pm source for more examples.

To forward all tcp traffic destined to port 80 to port 3000 instead (this is
useful for transparently forwarding traffic to a cache):

  use IPChains;

  $fw = IPChains->new(Source       => "0.0.0.0/0",
                      Destination  => "0.0.0.0/0",
                      DestPort     => "80",
                      Interface    => "eth0",
                      Rule         => "REDIRECT",
                      RedirectPort => "3000"
                      );
  $fw->append("input");

=head1 BUGS

Much of this is highly untested. Masquerading timeout setting and negative
attributes (!) aren't yet implemented.
Much of what's planned to be done hasn't been yet. This is
to be considered nothing more than an early beta to work out bugs in the
basic code, and get feedback on usefulness and improvements that could
be made.

=head1 AUTHOR

Jessica Hutchison (j@splatlabs.com). Please feel free to email me with
feedback, questions, or comments (or indeed patches/additions).

=head1 COPYRIGHT

This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself, with the exception
of the libipfwc.c, ipchains.c, and the files in include/ which 
have separate terms derived from those of the original ipchains 
sources. See COPYING for details of this license. Please see
README.ipchains for the README that was included with the
original source code for ipchains and contains copyrights and
credits for such. 
